File: testsuite/rdf.py
#
# Copyright (C) 2017 The ESPResSo project
#
# This file is part of ESPResSo.
#
# ESPResSo is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ESPResSo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

from __future__ import print_function
import unittest as ut
import espressomd
import numpy as np

class RdfTest(ut.TestCase):
    s = espressomd.System(box_l=[1.0, 1.0, 1.0])
    s.seed = s.cell_system.get_state()['n_nodes'] * [1234]

    def setUp(self):
        self.s.box_l = 3 * [10]
        self.s.part.clear()

    def bin_volumes(self, midpoints):
        bin_size = midpoints[1] - midpoints[0]
        r = midpoints - 0.5 * bin_size
        r2 = np.power(r, 2)

        # Volumes of the bins
        return 4.*3.14159*(r2*bin_size + r * bin_size**2 + bin_size**3/3.)

    def test_single_type(self):
        s = self.s

        n_part = 99
        dx = self.s.box_l[0] / float(n_part + 1)

        for i in range(n_part):
            s.part.add(id=i, pos=[i*dx, 0.5*s.box_l[1], 0.5*s.box_l[2]], type=0)

        r_bins = 50
        r_min = 0.5 * dx
        r_max = r_bins * dx
        rdf = s.analysis.rdf(rdf_type='rdf', type_list_a=[0,1],
                             r_min=r_min, r_max=r_max, r_bins=r_bins)
        rv = self.bin_volumes(rdf[0])
        rho = n_part / (s.box_l[0]**3)

        parts_in_bin = rdf[1] * rv * rho

        # All but the last bin should contain 2 particles
        self.assertTrue(np.allclose(parts_in_bin[:-1], 2.0))

    def test_mixed(self):
        s = self.s

        n_part = 99
        dx = self.s.box_l[0] / float(n_part + 1)

        for i in range(n_part):
            s.part.add(id=i, pos=[i*dx, 0.5*s.box_l[1], 0.5*s.box_l[2]], type=(i % 2))

        r_bins = 50
        r_min = 0.5 * dx
        r_max = r_bins * dx
        rdf01 = s.analysis.rdf(rdf_type='rdf', type_list_a=[0], type_list_b=[1],
                             r_min=r_min, r_max=r_max, r_bins=r_bins)
        rv = self.bin_volumes(rdf01[0])
        rho = 0.5 * n_part / (s.box_l[0]**3)

        parts_in_bin = rdf01[1] * rv * rho

        # Every odd bin should contain two parts
        self.assertTrue(np.allclose(parts_in_bin[0:-1:2], 2.0, rtol=1e-1))
        # Every odd bin should contain zero parts
        self.assertTrue(np.allclose(parts_in_bin[1:-1:2], 0.0))

        # Check symmetry
        rdf10 = s.analysis.rdf(rdf_type='rdf', type_list_a=[1], type_list_b=[0],
                             r_min=r_min, r_max=r_max, r_bins=r_bins)

        self.assertTrue(np.allclose(rdf10, rdf01))

    def test_av(self):
        s = self.s

        for i in range(200):
            s.part.add(id = i, pos=s.box_l * np.random.random(3), type=(i % 3))

        r_bins = 50
        r_min = 0.0
        r_max = 0.49 * s.box_l[0]
        rdf = s.analysis.rdf(rdf_type='rdf', type_list_a=[0,1,2],
                             r_min=r_min, r_max=r_max, r_bins=r_bins)

        for i in range(10):
            s.analysis.append()

        rdf_av = s.analysis.rdf(rdf_type='<rdf>', type_list_a=[0,1,2],
                             r_min=r_min, r_max=r_max, r_bins=r_bins)

        self.assertTrue(np.allclose(rdf[1], rdf_av[1]))

if __name__ == "__main__":
    ut.main()

File: src/core/observables/RDF.cpp
/*
 * Copyright (C) 2010-2022 The ESPResSo project
 *
 * This file is part of ESPResSo.
 *
 * ESPResSo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ESPResSo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "RDF.hpp"

#include "BoxGeometry.hpp"
#include "fetch_particles.hpp"
#include "system/System.hpp"

#include <utils/Vector.hpp>
#include <utils/for_each_pair.hpp>
#include <utils/math/int_pow.hpp>

#include <boost/mpi/communicator.hpp>
#include <boost/range/combine.hpp>

#include <cmath>
#include <cstddef>
#include <numbers>
#include <vector>

namespace Observables {
std::vector<double>
RDF::operator()(boost::mpi::communicator const &comm) const {
  auto const &local_particles_1 = fetch_particles(ids1());

  if (ids2().empty()) {
    return this->evaluate(comm, local_particles_1, {}, {});
  }

  auto const &local_particles_2 = fetch_particles(ids2());

  return this->evaluate(comm, local_particles_1, local_particles_2, {});
}

std::vector<double>
RDF::evaluate(boost::mpi::communicator const &comm,
              ParticleReferenceRange const &local_particles_1,
              ParticleReferenceRange const &local_particles_2,
              const ParticleObservables::traits<Particle> &traits) const {
  auto const positions_1 = detail::get_all_particle_positions(
      comm, local_particles_1, ids1(), traits, false);
  auto const positions_2 = detail::get_all_particle_positions(
      comm, local_particles_2, ids2(), traits, false);

  if (comm.rank() != 0) {
    return {};
  }

  auto const &box_geo = *System::get_system().box_geo;
  auto const bin_width = (max_r - min_r) / static_cast<double>(n_r_bins);
  auto const inv_bin_width = 1.0 / bin_width;
  std::vector<double> res(n_r_bins, 0.0);
  long int cnt = 0;
  auto op = [this, inv_bin_width, &cnt, &res, &box_geo](auto const &pos1,
                                                        auto const &pos2) {
    auto const dist = box_geo.get_mi_vector(pos1, pos2).norm();
    if (dist > min_r && dist < max_r) {
      auto const ind =
          static_cast<int>(std::floor((dist - min_r) * inv_bin_width));
      res[ind]++;
    }
    cnt++;
  };

  if (local_particles_2.empty()) {
    Utils::for_each_pair(positions_1, op);
  } else {
    auto const combine_1 = boost::combine(ids1(), positions_1);
    auto const combine_2 = boost::combine(ids2(), positions_2);

    auto op2 = [&op](auto const &it1, auto const &it2) {
      auto const &[id1, pos1] = it1;
      auto const &[id2, pos2] = it2;

      op(pos1, pos2);
    };

    auto cmp = [](auto const &it1, auto const &it2) {
      auto const &[id1, pos1] = it1;
      auto const &[id2, pos2] = it2;

      return id1 != id2;
    };
    Utils::for_each_cartesian_pair_if(combine_1, combine_2, op2, cmp);
  }
  if (cnt == 0)
    return res;
  // normalization
  auto const volume = box_geo.volume();
  for (std::size_t i = 0u; i < res.size(); ++i) {
    auto const r_in = static_cast<double>(i) * bin_width + min_r;
    auto const r_out = r_in + bin_width;
    auto const bin_volume =
        (4. / 3.) * std::numbers::pi *
        (Utils::int_pow<3>(r_out) - Utils::int_pow<3>(r_in));
    res[i] *= volume / (bin_volume * static_cast<double>(cnt));
  }

  return res;
}
} // namespace Observables
